### PR TITLE
[WIP] Weak Order 2.0 SRK methods #182

### PR DESCRIPTION
**Pull request by frankschae**
I added the constant cache part of the DRI1 method from the newer Ito paper.
I don't obtain the second order convergence on the stochastic part. 
The deterministic part converges with order 3 as expected. The associated test is implemented in ode_convergence_regression.jl

### CODE REVIEW

**Comment by ChrisRackauckas** [comment]
```julia
using Distributions
quantile(Normal(),1/6) # -0.9674215661017014
```

Declare this a constant and use the following:

```julia
dW_scaled = W.dW / sqrt(dt)
if _dW < -NORMAL_ONESIX_QUANTILE
  _dW = -sqrt(3dt)
elseif _dW > NORMAL_ONESIX_QUANTILE
  _dW = sqrt(3dt)
else
  _dW = zero(dt)
end
```

but in a `map` over the values.

**Comment by ChrisRackauckas** [comment]
Choose +- integrator.sqdt based on `sign(W.dZ)` and make this a process with two Brownians.

File: src/perform_step/srk_weak.jl [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+    H12 = [uprev + a121*k1*dt .+ b121*g1[k]*integrator.sqdt for k=1:m]
+  else
+    H12 = [uprev + a121*k1*dt + b121*g1[:,k]*integrator.sqdt for k=1:m]
+  end
+
+  # # H_i^(k), stage 3
+  if typeof(W.dW) <: Number
+    H13 = uprev + a131*k1*dt + b131*g1*integrator.sqdt
+  elseif is_diagonal_noise(integrator.sol.prob)
+    H13 = [uprev + a131*k1*dt .+ b131*g1[k]*integrator.sqdt for k=1:m]
```
Comment (ChrisRackauckas) [inline src/perform_step/srk_weak.jl]: these comprehensions should be broadcasts.

File: src/alg_utils.jl [hunk 2]
```diff
@@ -57,6 +57,9 @@ alg_order(alg::SRA2) = 2//1
 alg_order(alg::SRA3) = 2//1
 alg_order(alg::SOSRA) = 2//1
 alg_order(alg::SOSRA2) = 2//1
+
+alg_order(alg::DRI1) = 3//2 # deterministic // weak
```
Comment (ChrisRackauckas) [inline src/alg_utils.jl]: should be 1?

**Comment by ChrisRackauckas** [comment]
Check the order conditions in terms of the beta parts. My guess is that one of those are actually incorrect if it's not converging below 0.01.

**Comment by frankschae** [comment]
[EM-like.pdf](https://github.com/JuliaDiffEq/StochasticDiffEq.jl/files/4268033/EM-like.pdf)
The EM test look better now with the new sampling you proposed. I have not checked the conditions for the betas yet but what I found a bit odd is that also for the RI1 tableau (from the old ito paper) I don't find convergence.

**Comment by frankschae** [comment]
> [EM-like.pdf](https://github.com/JuliaDiffEq/StochasticDiffEq.jl/files/4268033/EM-like.pdf)
> The EM test look better now with the new sampling you proposed. I have not checked the conditions for the betas yet but what I found a bit odd is that also for the RI1 tableau (from the old ito paper) I don't find convergence.

**Comment by ChrisRackauckas** [comment]
Note that the other method is called the Simplified Euler-Maruyama scheme (SimplifiedEM()), and would be good to add as well. It should weakly converge with order 1.

http://www.diva-portal.se/smash/get/diva2:861891/FULLTEXT01.pdf

and it originates from KPS.

File: src/caches/srk_weak_caches.jl [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+
+  #beta41::T
+  beta42::T
+  beta43::T
+
+  #quantile(Normal(),1/6)
+  NORMAL_ONESIX_QUANTILE::T
+end
+
+function DRI1ConstantCache(T::Type, T2::Type)
```
Comment (ChrisRackauckas) [inline src/caches/srk_weak_caches.jl:68]:
```suggestion
function DRI1ConstantCache(::Type{T}, ::Type{T2}) where {T1,T2}
```

**Comment by ChrisRackauckas** [comment]
can you share the current long convergence plots here?

**Comment by YingboMa** [comment]
My inference fix is here https://github.com/SciML/StochasticDiffEq.jl/commit/ba3fa77f928554c9a79b2abd4cfdcdddcf87a645

**Comment by frankschae** [comment]
[DRI1-CPU-final-500000.0.pdf](https://github.com/SciML/StochasticDiffEq.jl/files/4423700/DRI1-CPU-final-500000.0.pdf)
[RI1-CPU-final-500000.0.pdf](https://github.com/SciML/StochasticDiffEq.jl/files/4423702/RI1-CPU-final-500000.0.pdf)

**Comment by frankschae** [comment]
The plots above  for DRI1 and RI1 are generated using the new file "srk_weak_final.jl"

**Comment by ChrisRackauckas** [comment]
That new test script isn't being ran? How long does it take to run?

**Comment by frankschae** [comment]
Only a few minutes..
The @ times of the script print:
28.555001 seconds (778.67 M allocations: 80.634 GiB, 34.83% gc time)
29.302624 seconds (841.61 M allocations: 81.294 GiB, 36.58% gc time)

**Comment by ChrisRackauckas** [comment]
Awesome, add it to the runtests and let's mark this as not WIP but done. We can continue to improve it later.

**Comment by ChrisRackauckas** [comment]
Awesome. Please open an issue on the inference bug so we can have the compiler team work on that. Let's continue refining the tests next, and add an in-place version.

**Pull request merged**